### PR TITLE
The enhanced BitmapScanner

### PR DESCRIPTION
**Pull request by lee-lei**
## What changes were proposed in this pull request?

The enhancements for bitmap scanner are below:

    1. Refactoring BitMapScanner and abstract the common BitmapReader
    2. Add BitmapReaderV1 based on the original BitMap Scanner for backward compatibility
    3. Add BitmapReaderV2 to directly get row Id list from fiber caches

## How was this patch tested?

mvn clean test package and all the tests passed.

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala [hunk 1]
```diff
@@ -16,502 +16,67 @@
...
-      initDesiredRowIdIterator()
-    } finally {
-      clearCache()
-    }
-
+    val fin = idxPath.getFileSystem(conf).open(idxPath)
+    val indexVersion = readHead(fin, 0)
+    val reader =
+      indexVersion match {
+        case 1 =>
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: please constant variable.

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManagerSuite.scala [hunk 2]
```diff
@@ -279,15 +279,15 @@ class FiberCacheManagerSuite extends SharedOapContext {
...
+  ignore("LRU blocks memory free") {
     val memorySizeInMB = (MemoryManager.cacheMemory / mbSize).toInt
     val dataInUse = generateData(mbSize)
     val fiberInUse = TestFiber(
       () => MemoryManager.toDataFiberCache(dataInUse), s"test fiber #${newFiberGroup}.0")
 
     // Put into cache and make it use
     val fiberCacheInUse = FiberCacheManager.get(fiberInUse, configuration)
-    assert(FiberCacheManager.pendingCount == 0)
+    // assert(FiberCacheManager.pendingCount == 0)
```
[the remark this answers is not in the document]
Comment (lee-lei) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManagerSuite.scala]: Sure.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala (hunk 1 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: Sure. Thanks for the good catch!

**Comment by lee-lei** [comment]
The disabled three fiber manager test cases are turned on by default. Thanks.

**Comment by adrian-wang** [comment]
Sorry for the conflicts, can you update this?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala (hunk 1 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: @lee-lei any update on this review?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      if (bmWfcSeq.nonEmpty) {
+        val iterator = BitmapUtils.iterator(bmWfcSeq)
+        bmRowIdIterator =
+          if (internalLimit > 0) iterator.take(internalLimit) else iterator
+        empty = false
+      } else {
+        empty = true
+      }
+    } finally {
+      clearCache
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]: add todo here to indicate this problem and file a issue.

[the remark this answers is not in the document]
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]: Sure.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReader.scala [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    var curOffset = 0
+    (0 until bmUniqueKeyListCount).map( idx => {
+      val (value, length) =
+        nnkr.readKey(data, curOffset)
+      curOffset += length
+      value
+    })
+  }
+
+  protected def getDesiredSegments(fin: FSDataInputStream): Unit = {
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReader.scala]: Also this

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV1.scala [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        } else {
+          bmRowIdIterator =
+            bitmapArray.reduceLeft(FastAggregation.or(_, _)).iterator.asScala
+        }
+        empty = false
+      } else {
+        empty = true
+      }
+    } finally {
+      clearCache
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV1.scala]: :nit if a function has side effect, use `()`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]: This approach will make it unusable I'm afraid

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  @transient private var bmRowIdIterator: Iterator[Int] = _
+  private var bmWfcSeq: Seq[OapBitmapWrappedFiberCache] = _
+  private var empty: Boolean = _
+
+  override def hasNext: Boolean = !empty && bmRowIdIterator.hasNext
+  override def next(): Int = bmRowIdIterator.next()
+  override def toString: String = "BitmapReaderV2"
+
+  override def clearCache(): Unit = {
+    super.clearCache
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]: `super.clearCache()`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReader.scala [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+  @transient protected var ordering: Ordering[Key] =
+    GenerateOrdering.create(keySchema)
+  @transient
+  protected lazy val nnkr: NonNullKeyReader = new NonNullKeyReader(keySchema)
+
+  protected val fiberCacheManager = OapRuntime.getOrCreate.fiberCacheManager
+
+  protected val BITMAP_FOOTER_SIZE = 6 * 8
+
+  protected var bmUniqueKeyListCount: Int = _
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReader.scala]: actually all the `bm` can be removed due to this is in a `Bitmapxxx`. 
Also, try not to use abbreviations.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]: We can record it and add a hook, when Guava decided to release the bitmap iterator, we release the cache related to `RoaringBitmap`, @lee-lei @linhong-intel

**Comment by lee-lei** [comment]
The existing guava cache has the listener to be called back when the cache entry is evicted. That is not a good approach to use call back everywhere the fiber cache is used. Thanks.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      if (bmWfcSeq.nonEmpty) {
+        val iterator = BitmapUtils.iterator(bmWfcSeq)
+        bmRowIdIterator =
+          if (internalLimit > 0) iterator.take(internalLimit) else iterator
+        empty = false
+      } else {
+        empty = true
+      }
+    } finally {
+      clearCache()
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]: add todo here to indicate current problem and file a issue.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala [hunk 9]
```diff
@@ -86,19 +91,15 @@ private[oap] class BitmapReaderV2(
...
-        empty = true
-      }
-    } finally {
-      clearCache()
+    initDesiredSegments()
+    bmWfcSeq = getDesiredWfcSeq
+    if (bmWfcSeq.nonEmpty) {
+      val iterator = BitmapUtils.iterator(bmWfcSeq)
+      bmRowIdIterator =
+        if (internalLimit > 0) iterator.take(internalLimit) else iterator
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]: :nit Can be in 1 line

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala [hunk 10]
```diff
@@ -47,7 +39,20 @@ private[oap] class BitmapReaderV2(
...
+ /* V2 is directly using fiber cache. Thus it needs to ensure that the bitmap fiber cache is
+  * residing in cache manager before the current query is finished. The current solution is to
+  * release the fibers after iterating is finished. However, the side effect is to keep the high
+  * memory pressure in cache manager.
+  * TODO: The solution for the above side effect is to improve the fiber cache manager to free
+  * unused cache memory to alleviate the cache eviction by size, and meanwhile keep still using
+  * cache memory. Even if it's evicted, put it back into cache as long as it's still using.
+  */
+  override def hasNext: Boolean =
+    if (!empty && bmRowIdIterator.hasNext) true else {
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapReaderV2.scala]:
```
if (...) {
true
} else {
...
}
```

**Pull request merged**